### ISSUE TITLE
Bug: [no-unnecessary-type-parameters] double false positive

### ISSUE THREAD

**Issue by abrahamguo** [typescript-eslint/typescript-eslint#9524]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=KYDwDg9gTgLgBAYwgOwM7wLYEMwHkBGAVnALxwA8A0nKDMMgCapzpQCWyA5gDRwBqvACoA%2BABQQiALjgBvOAG0A1sACecDnEoBdAPzS%2BcAL68AZsmmjlK6ZV4A3LABt9ASlLC4gtyWEAoOHAEhMAIMAB09DDswKjiRC5h2GCikVBqPnBmomE5qWpYzPK2-FouLgDcQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=&tokens=false

### Repro Code

```TypeScript
export const mapObj = <K extends string, V, T>(obj: { [key in K]?: V }, fn: (key: K, val: V) => T) =>
  Object.entries(obj).map(entry => fn(...entry as [K, V]));
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-parameters": "error",
  },
};
```

### Expected Result

The rule should not report on `V` — `V` is used 3 times

### Actual Result

The rule claims that `V` is only used once, when it's actually used 3 times.

**Comment by bradzacher** [comment]
The message is bad but the report is _kind of_ correct.

You can rewrite this to remove a parameter
```ts
export const mapObj =
  <Obj extends object, T>(
    obj: Obj,
    fn: (key: keyof Obj, val: Obj[keyof Obj]) => T,
  ) =>
    Object.entries(obj)
      .map(entry => fn(...entry as [keyof Obj, Obj[keyof Obj]]));

mapObj(
  { a: "", b: 1 },
  (
    k,
 // ^? "a" | "b"
    v
 // ^? string | number
  ) => true
)
```
[playground](https://www.typescriptlang.org/play/?#code/KYDwDg9gTgLgBAYwgOwM7wLYEMwHkBGAVnALwBQccAPAcaDMMgCapwRHAIwA0cAKgD4AFBUpsiALji1uoygDNkUoQGtgATylr1EedKK8AblgA2U2gG1tu-YQC6ASlID+syk5IC5tzjAB0jDBQAJbAqELshA7elH7YYEKBUOrOcIpCfplJKVisVho2Mrb5Onq0do4OANxkZPG0IpQA3nBYUgBE7bz4UgCMcAC+bnCNYipuAPQTcAB6APzehhRTswvuqUEArsBkDkA)

**Comment by danvk** [comment]
While the type parameter may not strictly be needed, this rule is nowhere near smart enough to figure that out. :) The issue is that `visitType` isn't descending into all parts of a mapped type. Fix coming up…

**Comment by abrahamguo** [comment]
@danvk since the rule reports that the type parameter is only used once, but it is actually used three times, does that mean that _two_ occurences of the type parameter have been missed?

If one missed occurrence was in the mapped type, what is the other missed occurrence?

**Comment by danvk** [comment]
@abrahamguo this is a reduced version of your code ([playground]):

```ts
declare function mapObj<
  K extends string,
  V,  // Type parameter V is used only once
>(
  obj: { [key in K]: V },
  fn: (key: K, val: V) => number,
): number[];
```

I've removed the function implementation because it's not relevant to this rule. `V` is declared once and used twice, so it appears a total of three times. So this is considered valid with my change, but not valid if you drop the use in `{ [key in K]: V }`.

[playground]: https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=CYUwxgNghgTiAEAzArgOzAFwJYHtXwFsoAHAeQCMArAHgCh54BpeEADwxFWAGd5uMYWVAHMANPXgA1UQwD0s%2BABUAnsQTFYUAiA4wp8LL2TcQweHgjLz6ELQB8ACgk4qALngBveAG0A1iCshJgBdd0l4AF9xBkRUdwd-ZXdGGQA3KAgwgEp4AF47eFRkAnIQGHEs9yKSsu9ggG5aIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWsRFCAtonyJoqDJCXQO0SODABfELqA&tsconfig=&tokens=false

**Comment by abrahamguo** [comment]
Ah. So the usage of `V` in the type assertion in the function implementation body (`as [K, V]`) doesn't count as a usage?

**Comment by danvk** [comment]
No. Type parameters are used to relate types in a type signature. If they only appear once in the type signature, then they're not relating anything. The implementation of a function can only affect its type signature if it has an inferred return type, which this rule specifically checks.
